### PR TITLE
fix(ext/node): MessagePort works

### PR DESCRIPTION
**Pull request by satyarohith**
Work in progress.

TODO:
- [ ] fix the order of events generated
- [x] add a test

Closes https://github.com/denoland/deno/issues/22951
Closes https://github.com/denoland/deno/issues/23001

### CODE REVIEW

**Comment by bartlomieju** [comment]
@satyarohith do you need any help with landing this PR? Would be great if we could land it before v1.42 this week.

**Comment by satyarohith** [comment]
> @satyarohith do you need any help with landing this PR? Would be great if we could land it before v1.42 this week.

Yes. I'm trying to figure out how close event would be emitted when either of the ports get closed.

**Comment by bartlomieju** [comment]
> > @satyarohith do you need any help with landing this PR? Would be great if we could land it before v1.42 this week.
> 
> Yes. I'm trying to figure out how close event would be emitted when either of the ports get closed.

I think you can override the `MessagePort::close` to emit the proper event?

**Comment by bartlomieju** [comment]
Additionally you can emit the event in `MessagePort::start` method when the `Interrupted` error is thrown (that tells that the other side has been closed). 

To do that you probably want to add a private API that allows to call a callback in either of these situations - you'd install the callback if the `MessagePort` is created from the Node polyfills.

**Comment by satyarohith** [comment]
This PR is working but the order of events fired is not similar to node:

```
➜  tst git:(fix_message_port_node_compat) ✗ node a.mjs        
worker: Hello from worker on parentPort!
mainPort: Hello from worker on workerPort!
worker port closed
mainPort closed
➜  tst git:(fix_message_port_node_compat) ✗ denod run -A a.mjs
worker: Hello from worker on parentPort!
worker port closed
mainPort: Hello from worker on workerPort!
mainPort closed
```

And according to https://nodejs.org/api/worker_threads.html#portclose, we need to emit "close" event on both the ports of a channel when either of the port is closed which is not happening right now.

File: ext/node/polyfills/worker_threads.ts [hunk 1]
```diff
@@ -424,10 +424,21 @@ export function receiveMessageOnPort(port: MessagePort): object | undefined {
   return { message: deserializeJsMessageData(data)[0] };
 }
 
+class NodeMessageChannel {
+  port1: MessagePort;
+  port2: MessagePort;
+
+  constructor() {
+    const { port1, port2 } = new MessageChannel(true);
```
Comment (satyarohith) [inline ext/node/polyfills/worker_threads.ts:432]: Not happy that nodeWorker flag needs to be public for MessageChannel but I'm not aware of a better choice here. Is there an alternative?

File: ext/web/13_message_port.js [hunk 2]
```diff
@@ -366,6 +401,7 @@ webidl.converters.StructuredSerializeOptions = webidl
   );
 
 function structuredClone(value, options) {
+  console.log("structuredClone invoked");
```
Comment (satyarohith) [inline ext/web/13_message_port.js:404]:
```suggestion
```
stale debug log

File: ext/web/13_message_port.js [hunk 3]
```diff
@@ -85,16 +86,37 @@ const MessageChannelPrototype = MessageChannel.prototype;
...
 
 /**
  * @param {number} id
  * @returns {MessagePort}
  */
-function createMessagePort(id) {
+function createMessagePort(id, nodeWorker = false) {
+  if (nodeWorker) {
+    defineEventHandler(MessagePort.prototype, "close");
+  }
```
Comment (bartlomieju) [inline ext/web/13_message_port.js:99]: We should remove this - we don't need `onclose` handler - just emitting `close` event for Node compat. Can be achieved without this.

File: ext/web/13_message_port.js [hunk 4]
```diff
@@ -85,16 +86,37 @@ const MessageChannelPrototype = MessageChannel.prototype;
...
+      } else if (name == "messageerror") {
+        port.onmessageerror = _listener;
+      } else if (name == "close") {
+        port.onclose = _listener;
+      } else {
+        throw new Error("Unsupported event: " + name);
+      }
+      return port;
+    };
+  }
```
Comment (bartlomieju) [inline ext/web/13_message_port.js:119]: I think all of this can be done in `node:worker_threads` similarly to what we do for `parentPort`

File: ext/web/13_message_port.js [hunk 5]
```diff
@@ -160,12 +184,19 @@ class MessagePort extends EventTarget {
           );
         } catch (err) {
           if (ObjectPrototypeIsPrototypeOf(InterruptedPrototype, err)) break;
+          if (this[_nodeWorker]) this.dispatchEvent(new Event("close"));
```
Comment (bartlomieju) [inline ext/web/13_message_port.js:187]: Now this is tricky and I'm not yet sure how to handle that. Maybe we keep `this[_nodeWorker]` but make it non-enumerable and then we set it to true only from `node:worker_threads`

File: ext/web/13_message_port.js [hunk 6]
```diff
@@ -103,6 +125,8 @@ class MessagePort extends EventTarget {
   [_id] = null;
   /** @type {boolean} */
   [_enabled] = false;
+  [_nodeWorker] = false;
+  [_closeEventDispatched] = false;
```
Comment (bartlomieju) [inline ext/web/13_message_port.js:129]: We should make these two non-enumerable so that inspecting from user code doesn't show them

File: ext/web/13_message_port.js (hunk 5 above)
Comment (bartlomieju) [reply ext/web/13_message_port.js:187]: I'm still thinking that we could handle this by adding `this[_workerThreadCloseCb]` that would be set from `node:worker_threads` and it will be up to the callback to decide what to do (eg. `this.emit("close")`).

File: ext/web/13_message_port.js (hunk 3 above)
Comment (satyarohith) [reply ext/web/13_message_port.js:99]: node supports `on("close", handler)` on a worker so I introduced it - am I missing something?

File: ext/web/13_message_port.js (hunk 4 above)
Comment (satyarohith) [reply ext/web/13_message_port.js:119]: Ah, I'll try that. I think I had to use a flag in web implementation because deserialization of workerData didn't setup the MessagePort class correctly in worker thread.

File: ext/web/13_message_port.js (hunk 3 above)
Comment (bartlomieju) [reply ext/web/13_message_port.js:99]: `onclose` is "Web" way, while `on("close", ..)` is the Node way. If you do it this way, then suddenly `MessagePort.onclose` becomes a thing, which is not available in the https://developer.mozilla.org/en-US/docs/Web/API/MessagePort#events

File: ext/web/13_message_port.js (hunk 5 above)
Comment (satyarohith) [reply ext/web/13_message_port.js:187]: Sure. A callback should be fine too. I'll look into it.

File: ext/node/polyfills/worker_threads.ts [hunk 7]
```diff
@@ -424,10 +432,42 @@ export function receiveMessageOnPort(port: MessagePort): object | undefined {
...
+  port2: MessagePort;
+
+  constructor() {
+    const { port1, port2 } = new MessageChannel();
+    this.port1 = patchPort(port1);
+    this.port2 = patchPort(port2);
+  }
+}
+
+function patchPort(port: MessagePort) {
```
Comment (bartlomieju) [inline ext/node/polyfills/worker_threads.ts:446]: Nitpick: let's rename this one to `webMessagePortToNodeMessagePort`

File: ext/node/polyfills/worker_threads.ts [hunk 8]
```diff
@@ -424,10 +432,42 @@ export function receiveMessageOnPort(port: MessagePort): object | undefined {
...
+    } else if (name == "messageerror") {
+      port.onmessageerror = _listener;
+    } else if (name == "close") {
+      port.addEventListener("close", _listener);
+    } else {
+      throw new Error(`Unknown event: "${name}"`);
+    }
+    return this;
+  };
+  port[workerThreadCloseCb] = () => {
```
Comment (bartlomieju) [inline ext/node/polyfills/worker_threads.ts:461]: Great, that's what I had in mind

File: ext/web/13_message_port.js [hunk 9]
```diff
@@ -85,6 +86,8 @@ const MessageChannelPrototype = MessageChannel.prototype;
 const _id = Symbol("id");
 const MessagePortIdSymbol = _id;
 const _enabled = Symbol("enabled");
+const workerThreadCloseCb = Symbol("workerThreadCloseCb");
```
Comment (bartlomieju) [inline ext/web/13_message_port.js:89]:
```suggestion
const nodeWorkerThreadCloseCb = Symbol("nodeWorkerThreadCloseCb");
```

File: ext/web/13_message_port.js [hunk 10]
```diff
@@ -85,6 +86,8 @@ const MessageChannelPrototype = MessageChannel.prototype;
...
+const _workerThreadCloseCbInvoked = Symbol("workerThreadCloseCbInvoked");
```
Comment (bartlomieju) [inline ext/web/13_message_port.js:90]:
```suggestion
const nodeWorkerThreadCloseCbInvoked = Symbol("nodeWorkerThreadCloseCbInvoked");
```

File: ext/web/13_message_port.js [hunk 11]
```diff
@@ -160,9 +171,25 @@ class MessagePort extends EventTarget {
           );
         } catch (err) {
           if (ObjectPrototypeIsPrototypeOf(InterruptedPrototype, err)) break;
+          if (
+            typeof this[workerThreadCloseCb] == "function" &&
+            !this[_workerThreadCloseCbInvoked]
+          ) {
+            this[workerThreadCloseCb]();
+            this[_workerThreadCloseCbInvoked] = true;
+          }
```
Comment (bartlomieju) [inline ext/web/13_message_port.js:180]: How about this move it to a top-level helper like: `nodeWorkerThreadMaybeInvokeCloseCb`?

File: tests/integration/worker_tests.rs [hunk 12]
```diff
@@ -125,3 +126,9 @@ itest!(node_worker_auto_exits {
   output: "workers/node_worker_auto_exits.mjs.out",
   exit_code: 0,
 });
+
+itest_flaky!(node_worker_message_port {
+  args: "run --quiet --allow-read workers/node_worker_message_port.mjs",
+  output: "workers/node_worker_message_port.mjs.out",
+  exit_code: 0,
+});
```
Comment (bartlomieju) [inline tests/integration/worker_tests.rs:134]: So it passes sometimes?

File: runtime/js/99_main.js [hunk 13]
```diff
@@ -924,7 +924,7 @@ function bootstrapWorkerRuntime(
     ObjectDefineProperty(globalThis, "Deno", core.propReadOnly(finalDenoNs));
 
     const workerMetadata = maybeWorkerMetadata
-      ? messagePort.deserializeJsMessageData(maybeWorkerMetadata)
+      ? messagePort.deserializeJsMessageData(maybeWorkerMetadata, true)
```
Comment (bartlomieju) [inline runtime/js/99_main.js:927]: I think this is not needed anymore?

File: ext/web/internal.d.ts
Comment (bartlomieju) [inline ext/web/internal.d.ts:1]: These changes also seems no longer needed

File: tests/unit_node/worker_threads_test.ts [hunk 14]
```diff
@@ -16,7 +16,7 @@ Deno.test("[node/worker_threads] BroadcastChannel is exported", () => {
 });
 
 Deno.test("[node/worker_threads] MessageChannel are MessagePort are exported", () => {
-  assertEquals<unknown>(workerThreads.MessageChannel, MessageChannel);
```
Comment (bartlomieju) [inline tests/unit_node/worker_threads_test.ts:19]: Why this change?

Comment (satyarohith) [reply tests/unit_node/worker_threads_test.ts:19]: Because the MessageChannel exported from worker_threads is now NodeMessageChannel

File: tests/integration/worker_tests.rs (hunk 12 above)
Comment (satyarohith) [reply tests/integration/worker_tests.rs:134]: I marked it as flaky because the order is not predictable. But it doesn't match the output of Node yet.

**Review by bartlomieju — APPROVED** [review]
LGTM, let's address the ordering problem in a separate PR.
